### DISCUSSION TITLE
Piping file from getServerSideProps?

### DISCUSSION THREAD

**Discussion by justin-harr** [vercel/next.js#30100] [Help]
I have PDFs stored a third party file server bucket, so an example PDF might be at: https://myfileserver.com/sample.pdf

I would like to be able to navigate to a Next page with a dynamic slug, use that slug to fetch the filename from a database, and fetch and pipe back the PDF as if you navigated to the PDF file directly, but without a redirect. For example this is how you would do it in Express:

```
const express = require('express')
const app = express()
const request = require('request');
app.get('/sample', (req, res) => {
  return request('https://myfileserver.com/sample.pdf').pipe(res);
})
app.listen(3000);
```
If you then navigated to http://localhost:3000/sample you would see the PDF as if you directly navigated to https://myfileserver.com/sample.pdf, but the URL wouldn't change you'd still be at http://localhost:3000/sample. Is there a good NextJs way to do this? I assume it would need to be sent from getServerSideProps since I don't want to have to prefix my URLs with /api.

**Comment by gaston-flores** [comment]
If you are using `next start` to run your app, you can use a [custom server](https://github.com/vercel/next.js/tree/canary/examples/custom-server-express) to do this in the exact way you describe here, but I don't think you can do this with vanilla next.js routing. As I understand it, outside of the `/api` directory every page has to be a React component: there's no way for a page to be just "`getServerSideProps`".

**Comment by ghost** [comment]
Also looking for a way to do this in Next.js 12.
In Next.js 11, I was able to serve a PDF from getServerSideProps using `context.res.end(pdfBuffer)` but this no longer works in Next.js 12.
Here is a demo repository: https://github.com/kelleg1/nextjs-pdf-demo

@timneutkens Is there a suggested method for this?

**Edit:** I suppose one way to serve the PDF is to output the file to the file system and then redirect to that file, however that does cause the URL displayed in the user's browser to change. I've updated the above repository with an example of doing that. I'm still wondering if there is a better way, if the previous functionality of Next.js 11 can be restored, or if the Next.js team might consider adding a PDF rendering feature.
